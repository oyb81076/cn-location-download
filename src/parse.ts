import { load } from "cheerio";
import { resolve } from "url";
import { IData, IOptions } from "./faces";

export const parse: IOptions["parse"] = ({ depth, url }, content) => {
  if (!url) { return []; }
  switch (depth) {
    case 0: return parseRoot(url, content);
    case 1: return parseOthers(url, content);
    case 2: return parseOthers(url, content);  // 有些地方返回的就直接是 county, 有些则是 towns
    case 3: return parseOthers(url, content);
    case 4: return parseOthers(url, content);
    default: throw new Error("错误的depth参数");
  }
};

export function parseRoot(url: string, content: string): IData[] {
  const $ = load(content);
  return $(".provincetr a").toArray().map((el): IData => {
    const href = el.attribs.href;
    const code = parseInt(href.split(".")[0], 10) * 100000000;
    const name = $(el).text();
    const link = resolve(url, href);
    return { name, url: link, code };
  });
}

export function parseOthers(url: string, content: string): IData[] {
  const $ = load(content);
  return $(".citytr,.countytr,.towntr,.villagetr").toArray().map((tr) => {
    const $tr = $(tr);
    const $children = $tr.children();
    const href = $tr.find("a").attr("href");
    const code = parseInt($children.first().text(), 10);
    const name = $children.last().text();
    if (href) {
      return { name, code, url: resolve(url, href) };
    } else {
      return { name, code };
    }
  });
}
