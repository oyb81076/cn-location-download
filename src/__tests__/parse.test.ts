import { readFileSync } from "fs";
import { join } from "path";
import { parseOthers } from "../parse";

test("442000106", () => {
  const HTML = readFileSync(join(__dirname, "./442000106.html"), { encoding: "utf8" });
  const URL = "http://www.stats.gov.cn/tjsj/tjbz/tjyqhdmhcxhfdm/2018/44/20/442000106.html";
  expect(parseOthers(URL, HTML)).toMatchSnapshot();
});
