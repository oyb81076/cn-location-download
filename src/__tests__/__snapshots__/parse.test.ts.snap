// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Object {
    "code": 442000106001,
    "name": "汇源社区居民委员会",
  },
  Object {
    "code": 442000106201,
    "name": "康乐村民委员会",
  },
  Object {
    "code": 442000106202,
    "name": "龙山村民委员会",
  },
  Object {
    "code": 442000106203,
    "name": "龙瑞村民委员会",
  },
  Object {
    "code": 442000106204,
    "name": "云汉村民委员会",
  },
  Object {
    "code": 442000106205,
    "name": "乐群村民委员会",
  },
  Object {
    "code": 442000106206,
    "name": "龙头环村民委员会",
  },
  Object {
    "code": 442000106207,
    "name": "圣狮村民委员会",
  },
  Object {
    "code": 442000106208,
    "name": "象角村民委员会",
  },
  Object {
    "code": 442000106209,
    "name": "沙溪村民委员会",
  },
  Object {
    "code": 442000106210,
    "name": "虎逊村民委员会",
  },
  Object {
    "code": 442000106211,
    "name": "濠涌村民委员会",
  },
  Object {
    "code": 442000106212,
    "name": "涌头村民委员会",
  },
  Object {
    "code": 442000106213,
    "name": "中兴村民委员会",
  },
  Object {
    "code": 442000106214,
    "name": "涌边村民委员会",
  },
  Object {
    "code": 442000106215,
    "name": "港园村民委员会",
  },
]
`;

exports[`442000106 1`] = `
Array [
  Object {
    "code": 442000106001,
    "name": "汇源社区居民委员会",
  },
  Object {
    "code": 442000106201,
    "name": "康乐村民委员会",
  },
  Object {
    "code": 442000106202,
    "name": "龙山村民委员会",
  },
  Object {
    "code": 442000106203,
    "name": "龙瑞村民委员会",
  },
  Object {
    "code": 442000106204,
    "name": "云汉村民委员会",
  },
  Object {
    "code": 442000106205,
    "name": "乐群村民委员会",
  },
  Object {
    "code": 442000106206,
    "name": "龙头环村民委员会",
  },
  Object {
    "code": 442000106207,
    "name": "圣狮村民委员会",
  },
  Object {
    "code": 442000106208,
    "name": "象角村民委员会",
  },
  Object {
    "code": 442000106209,
    "name": "沙溪村民委员会",
  },
  Object {
    "code": 442000106210,
    "name": "虎逊村民委员会",
  },
  Object {
    "code": 442000106211,
    "name": "濠涌村民委员会",
  },
  Object {
    "code": 442000106212,
    "name": "涌头村民委员会",
  },
  Object {
    "code": 442000106213,
    "name": "中兴村民委员会",
  },
  Object {
    "code": 442000106214,
    "name": "涌边村民委员会",
  },
  Object {
    "code": 442000106215,
    "name": "港园村民委员会",
  },
]
`;
